app/models/robot_world.rb

```diff
@@ -1,56 +1,80 @@
+    @database = database
+    @robot_ages = {"years" => [], "months" => [], "days" => []}
+    @hires_by_year = {}
   end
 
   def create(robot)
-    world.transaction do
-      world['robots'] ||= []
-      world['population'] ||= 0
-      world['population'] += 1
-      world['robots'] << new_robot(world['population'], robot[:name], robot[:city], robot[:state],
-                                   robot[:birthdate], robot[:date_hired], robot[:department], "https://robohash.org/#{world['population']}")
-    end
+    robo = {:name => robot["name"], :city => robot["city"], 
+                                 :state => robot["state"], :birthdate => format_date(robot["birthdate"]),
+                                 :date_hired => format_date(robot["date_hired"]), 
+                                 :department => robot["department"],
+                                 :avatar => "https://robohash.org/#{rand(100)}"}
+
+    database.execute("INSERT INTO robots (name, city, state, birthdate,
+                                        date_hired, department, avatar)
+                     VALUES (\"#{robo[:name]}\", \"#{robo[:city]}\", \"#{robo[:state]}\",
+                            \"#{robo[:birthdate]}\", \"#{robo[:date_hired]}\",
+                            \"#{robo[:department]}\", \"#{robo[:avatar]}\");"
+                    )
   end
 
-  Robot = Struct.new(:id, :name, :city, :state,
-                     :birthdate,
-                     :date_hired, :department, :avatar)
+  def new_robot(robot_details)
+    Robot.new(robot_details)
+  end
 
-  def new_robot(id, name, city, state, birthday, date_hired, department, avatar)
-    Robot.new(id, name, city, state, birthday, date_hired, department, avatar)
+  def format_date(date)
+    unless date.nil?
+      "#{date["month"].to_s.rjust(2,"0")}/#{date["day"].to_s.rjust(2,"0")}/#{date["year"].to_s.rjust(4,"0")}"
+    end
   end
 
   def all
-    world.transaction do
-      world['robots'] ||= []
-    end
+    database.execute("SELECT * FROM robots;")
   end
 
   def find(robot_id)
-    all.find{|robot| robot.id == robot_id}
-  end
-
-  def update(robot_id, robot_data)
-    world.transaction do
-      targeted_robot = world['robots'].find{|robot| robot.id == robot_id}
-      targeted_robot[:name] = robot_data[:name]
-      targeted_robot[:city] = robot_data[:city]
-      targeted_robot[:state] = robot_data[:state]
-      targeted_robot[:birthdate] = robot_data[:birthdate]
-      targeted_robot[:date_hired] = robot_data[:date_hired]
-      targeted_robot[:department] = robot_data[:department]
-      targeted_robot[:avatar] = robot_data[:avatar]
-    end
+    robot = new_robot(all.find do |robot| 
+      robot["id"] == robot_id
+    end.delete_if{|k,v| k.is_a?(Fixnum)})
+  end
+
+  def update(robot_id, new_robot_data)
+    database.execute("UPDATE robots 
+                     SET name=? 
+                     WHERE id=?;", [new_robot_data["name"], robot_id]) unless new_robot_data["name"].nil?
```

Comment:
if the field the user is updating is changed to nil, don't you still want to save that? perhaps they changed it to nil and/or an empty string on purpose?